src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,158 @@
+#include "partial_hash_index_impl.hpp"
+
+#include "storage/segment_iterate.hpp"
+
+namespace hyrise {
+
+template <typename DataType>
+PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id)
+    : BasePartialHashIndexImpl() {
+  insert_entries(chunks_to_index, column_id);
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::insert_entries(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id) {
+  auto indexed_chunks = size_t{0};
+  for (const auto& chunk : chunks_to_index) {
+    if (_indexed_chunk_ids.contains(chunk.first)) {
+      // Index already contains entries for the given chunk.
+      continue;
+    }
+
+    _indexed_chunk_ids.insert(chunk.first);
+    ++indexed_chunks;
+
+    // Iterate over the segment to index and populate the index.
+    const auto& indexed_segment = chunk.second->get_segment(column_id);
+    segment_iterate<DataType>(*indexed_segment, [&](const auto& position) {
+      const auto row_id = RowID{chunk.first, position.chunk_offset()};
+      // If value is NULL, add to NULL vector, otherwise add into value map.
+      if (position.is_null()) {
+        if (!_null_positions.contains(DataType{})) {
```

Comment:
`null_positions` is a map with at most one bucket, right? So `null_positions.empty()` could safe one construction and the lookup (if the map does not have a shortcut). Not sure if I asked that before: Does the sparse map conform to the STL regarding `operator[]`? If so, then the switch is useless:
> Returns a reference to the value that is mapped to a key equivalent to `key`, performing an insertion if such key does not already exist.

If I wanna be picky, I think @Bouncner's comment about the construction/copy of `row_id` is still valid.

Reply:
Good catch, in this case the map is conform to STL.
Do you know how to make Martins suggestion work?
RowID does not have a constructor emplace_back() where emplace_back() can pass the values to, and creating a constructor for RowID is not suitable because of the NULL_ROW_ID constexpression.
How do you like `_null_positions[DataType{}].emplace_back(RowID{chunk.first, position.chunk_offset()});`?

Reply:
> creating a constructor for RowID is not suitable because of the NULL_ROW_ID constexpression

Could you elaborate on that a bit further? Don't get the problem yet.

Reply:
If I've understood everything right, `.emplace_back()` forwards its parameters to the constructor of the class of the vector.
But since the `RowID` struct does not have a constructor which accepts two parameters, `.emplace_back(chunk.first, position.chunk_offset())` does not work.
I tried implementing a custom constructor for RowID, but I occurred some weird errors regarding some constexpressions in the types.hpp.

Reply:
You can checkout commit `542dcb8a8d6a470cabd430dc94d21e9361b0eb24` to see what fails.

Reply:
```diff
diff --git a/src/lib/types.hpp b/src/lib/types.hpp
index 6b818a5fb..b10c1f5d0 100644
--- a/src/lib/types.hpp
+++ b/src/lib/types.hpp
@@ -102,11 +102,12 @@ constexpr ChunkOffset INVALID_CHUNK_OFFSET{std::numeric_limits<ChunkOffset::base
 constexpr ChunkID INVALID_CHUNK_ID{std::numeric_limits<ChunkID::base_type>::max()};

 struct RowID {
-  RowID(const ChunkID init_chunk_id, const ChunkOffset init_chunk_offset) : chunk_id{init_chunk_id}, chunk_offset{init_chunk_offset} {}
+  constexpr RowID(const ChunkID init_chunk_id, const ChunkOffset init_chunk_offset) : chunk_id{init_chunk_id}, chunk_offset{init_chunk_offset} {}
+  RowID() = default;
   ChunkID chunk_id{INVALID_CHUNK_ID};
   ChunkOffset chunk_offset{INVALID_CHUNK_OFFSET};

-  // Faster than row_id == ROW_ID_NULL, since we only compare the ChunkOffset
+  // Faster than row_id == NULL_ROW_ID, since we only compare the ChunkOffset
   bool is_null() const {
     return chunk_offset == INVALID_CHUNK_OFFSET;
   }
```

Reply:
Ok, that was easy
Thank you a lot!